File: lib/src/geocode.dart
import 'dart:convert';
import 'dart:io';

/// One place found for a search — enough to jump the map near it.
/// [bounds] is the place's own extent (south, north, west, east) when
/// the service reports it: a street deserves street zoom, a country
/// country zoom.
class GeoHit {
  final String name;
  final double lat;
  final double lon;
  final (double, double, double, double)? bounds;

  const GeoHit(this.name, this.lat, this.lon, {this.bounds});
}

/// Signature the picker takes, so tests inject a stub instead of the
/// network.
typedef GeocodeSearch = Future<List<GeoHit>> Function(String query);

/// Address/city/country search via OSM Nominatim — besides map tiles,
/// the app's only outbound call (named in the privacy policy). Runs
/// only when the user submits a search in the position picker, on the
/// map when the catalog knows no such name, or when the flier scan is
/// asked to locate the owner's address; no reverse geocoding anywhere.
Future<List<GeoHit>> nominatimSearch(String query) async {
  final client = HttpClient()
    ..userAgent = 'catlog/0.2 (https://github.com/paxel/catlog)'
    ..connectionTimeout = const Duration(seconds: 10);
  try {
    final uri = Uri.https('nominatim.openstreetmap.org', '/search', {
      'q': query,
      'format': 'jsonv2',
      'limit': '8',
      // The address in parts, so it can be written the way the place
      // itself writes addresses — not in Nominatim's one order for all.
      'addressdetails': '1',
    });
    final req = await client.getUrl(uri);
    final res = await req.close();
    if (res.statusCode != HttpStatus.ok) return const [];
    final body = await utf8.decoder.bind(res).join();
    return [
      for (final hit in jsonDecode(body) as List)
        GeoHit(
          formatAddress((hit as Map)['address'],
              fallback: hit['display_name'] as String),
          double.parse(hit['lat'] as String),
          double.parse(hit['lon'] as String),
          bounds: _bounds(hit['boundingbox']),
        )
    ];
  } finally {
    client.close(force: true);
  }
}

/// Countries that write the house number before the street and the
/// postcode after the town ("12 Main Street, Springfield 62704").
/// Everywhere else the street comes first and the postcode before the
/// town ("Grimmaische Straße 12, 04109 Leipzig").
const _numberFirst = {
  'us', 'ca', 'gb', 'ie', 'fr', 'au', 'nz', 'za', 'in', 'sg', 'hk', 'my',
  'ph', 'lu',
};

/// A street address from Nominatim's `address` parts, in the order the
/// hit's own country uses. Hits without a street — a town, a region, a
/// country — keep [fallback], Nominatim's full name, which reads fine
/// for those. The country itself is left off: an address is looked up
/// where the cats live.
String formatAddress(Object? address, {required String fallback}) {
  if (address is! Map) return fallback;
  String? part(List<String> keys) {
    for (final k in keys) {
      final v = address[k];
      if (v is String && v.trim().isNotEmpty) return v.trim();
    }
    return null;
  }

  final street = part(['road', 'pedestrian', 'footway', 'path', 'square']);
  if (street == null) return fallback;
  final number = part(['house_number']);
  final place =
      part(['city', 'town', 'village', 'municipality', 'hamlet', 'suburb']);
  final postcode = part(['postcode']);
  final numberFirst =
      _numberFirst.contains(part(['country_code'])?.toLowerCase());
  final streetLine = number == null
      ? street
      : numberFirst
          ? '$number $street'
          : '$street $number';
  final placeLine = [
    if (numberFirst) ...[?place, ?postcode] else ...[?postcode, ?place],
  ].join(' ');
  return placeLine.isEmpty ? streetLine : '$streetLine, $placeLine';
}

/// Nominatim's `boundingbox`: four strings, ordered
/// [minlat, maxlat, minlon, maxlon]. Anything else is ignored.
(double, double, double, double)? _bounds(Object? raw) {
  if (raw is! List || raw.length != 4) return null;
  final v = [for (final x in raw) double.tryParse('$x')];
  if (v.any((x) => x == null)) return null;
  return (v[0]!, v[1]!, v[2]!, v[3]!);
}

File: desktop/crates/catlog-core/src/geocode.rs
//! Address, city and country search through OSM Nominatim: besides map
//! tiles the app's only outbound call, made only when the keeper
//! submits a search. No reverse geocoding anywhere.

/// One place found for a search. `bounds` is the place's own extent
/// (south, north, west, east) when the service reports it.
#[derive(Debug, Clone, PartialEq)]
pub struct GeoHit {
    pub name: String,
    pub lat: f64,
    pub lon: f64,
    pub bounds: Option<(f64, f64, f64, f64)>,
}

/// Where places are looked up: the service, or a stand-in in tests.
pub trait Geocoder: Send + Sync {
    fn search(&self, query: &str) -> Result<Vec<GeoHit>, String>;
}

pub struct Nominatim;

impl Geocoder for Nominatim {
    fn search(&self, query: &str) -> Result<Vec<GeoHit>, String> {
        let agent = ureq::Agent::config_builder()
            .user_agent(crate::tiles::TILE_USER_AGENT)
            .timeout_global(Some(std::time::Duration::from_secs(10)))
            .build()
            .new_agent();
        let url = format!(
            "https://nominatim.openstreetmap.org/search?q={}&format=jsonv2&limit=8&addressdetails=1",
            crate::registry::encode_component(query)
        );
        let text = agent
            .get(&url)
            .call()
            .map_err(|e| e.to_string())?
            .into_body()
            .read_to_string()
            .map_err(|e| e.to_string())?;
        parse_hits(&text)
    }
}

/// The hits in a Nominatim `jsonv2` answer.
pub fn parse_hits(json: &str) -> Result<Vec<GeoHit>, String> {
    let value: serde_json::Value = serde_json::from_str(json).map_err(|e| e.to_string())?;
    let list = value.as_array().ok_or("not a list")?;
    Ok(list
        .iter()
        .filter_map(|hit| {
            let lat: f64 = hit.get("lat")?.as_str()?.parse().ok()?;
            let lon: f64 = hit.get("lon")?.as_str()?.parse().ok()?;
            let name = hit.get("display_name")?.as_str()?.to_string();
            let bounds = hit.get("boundingbox").and_then(|b| {
                let b = b.as_array()?;
                let f = |i: usize| b.get(i)?.as_str()?.parse::<f64>().ok();
                Some((f(0)?, f(1)?, f(2)?, f(3)?))
            });
            Some(GeoHit {
                name,
                lat,
                lon,
                bounds,
            })
        })
        .collect())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn nominatim_answers_parse_into_hits() {
        let json = r#"[{"lat":"51.34","lon":"12.37","display_name":"Leipzig, Sachsen","boundingbox":["51.2","51.4","12.2","12.5"]},{"lat":"x"},{"lat":"1","lon":"2","display_name":"Somewhere"}]"#;
        let hits = parse_hits(json).unwrap();
        assert_eq!(hits.len(), 2);
        assert_eq!(hits[0].name, "Leipzig, Sachsen");
        assert_eq!(hits[0].bounds, Some((51.2, 51.4, 12.2, 12.5)));
        assert_eq!(hits[1].bounds, None);
        assert!(parse_hits("{}").is_err());
        assert!(parse_hits("junk").is_err());
    }
}
